source/grid/grid_tools.cc

```diff
@@ -2088,6 +2088,159 @@ namespace GridTools
+          BoundingBox < spacedim > bbox =
+            internal::BoundingBoxPredicate::compute_cell_predicate_bounding_box <MeshType> (cell, predicate, has_predicate);
+          if (has_predicate)
+            bounding_boxes.push_back(bbox);
+        }
+
+    // Creating a Bounding Box for all cells on the chosen refinement_level
+    for (typename MeshType::cell_iterator cell: mesh.cell_iterators_on_level(refinement_level))
+      {
+        bool has_predicate = false;
+        BoundingBox < spacedim > bbox =
+          internal::BoundingBoxPredicate::compute_cell_predicate_bounding_box <MeshType> (cell, predicate, has_predicate);
+        if (has_predicate)
+          bounding_boxes.push_back(bbox);
+      }
+
+
+    // Part 1: merging neighbours
+    // This array stores the indices of arrays we have already merged
+    std::vector<unsigned int> merged_boxes_idx;
+    bool found_neighbors = true;
+
+    // We merge only nighbours which can be expressed by a single bounding box
+    while (found_neighbors)
+      {
+        found_neighbors = false;
+        for (unsigned int i=0; i<bounding_boxes.size()-1; ++i)
+          {
+            if ( std::find(merged_boxes_idx.begin(),merged_boxes_idx.end(),i) == merged_boxes_idx.end())
+              for (unsigned int j=i+1; j<bounding_boxes.size(); ++j)
+                if ( std::find(merged_boxes_idx.begin(),merged_boxes_idx.end(),j) == merged_boxes_idx.end()
+                     && bounding_boxes[i].is_neighbor(bounding_boxes[j]) == mergeable_neighbors  )
+                  {
+                    bounding_boxes[i].merge_with(bounding_boxes[j]);
+                    merged_boxes_idx.push_back(j);
+                    found_neighbors = true;
+                  }
+          }
+      }
+
+    // Copying the merged boxes into merged_b_boxes
+    std::vector< BoundingBox < spacedim > > merged_b_boxes;
+    for (unsigned int i=0; i<bounding_boxes.size(); ++i)
+      if (std::find(merged_boxes_idx.begin(),merged_boxes_idx.end(),i) == merged_boxes_idx.end())
+        merged_b_boxes.push_back(bounding_boxes[i]);
+
+    // Part 2: if there are too many bounding boxes, merging smaller boxes
+    if (merged_b_boxes.size() > max_boxes)
+      {
+        std::vector<double> volumes;
+        for (unsigned int i=0; i< merged_b_boxes.size(); ++i)
+          volumes.push_back(merged_b_boxes[i].volume());
+
+        while ( merged_b_boxes.size() > max_boxes)
+          {
+            unsigned int min_idx = std::min_element(volumes.begin(),volumes.end()) -
+                                   volumes.begin();
+            volumes.erase(volumes.begin() + min_idx);
+            //Finding a neighbor
+            bool not_removed = true;
+            for (unsigned int i=0; i<merged_b_boxes.size() && not_removed; ++i)
+              if ( i != min_idx && merged_b_boxes[i].is_neighbor(merged_b_boxes[min_idx]) > 1 )
+                {
+                  merged_b_boxes[i].merge_with(merged_b_boxes[min_idx]);
```

Comment:
why do you need to merge here again (we already merged everything above or not?)?

Reply:
This is an extra round: in the first part we merge only if merging doesn't change the space you're describing with the bounding boxes ( for instance in 1d [0,1] and [1,2] can be described with [0,2] without losing anything).

Now that I think of it, this passage in 1d won't do anything. In higher dimension the idea is that, for instance:
if I have the 2d bounding box with low/left and high right points being (0,0),(1,1) and the bounding box (0,1),(eps,1+eps), with eps small, I might want (in my case I want to) merge them to (0,0) (1,1+eps): losing "accuracy" in the spatial description but reducing the number of cells I'm using.

What happens now is that, after this refinement, we might have more bounding boxes than we meant: what the algorithm is does is merging smaller bounding boxes. I want to do this only if the first algorithm didn't reduce enough the number of BBoxes.